### PR TITLE
Import watch only pubkeys to the keypool if private keys are disabled

### PR DESCRIPTION
**Pull request by achow101**
If the wallet has private keys disabled, allow importing public keys into the keypool. A `keypool` option has been added to `importmulti` in order to signal that the keys should be added to the keypool.

### CODE REVIEW

File: test/functional/wallet_importmulti.py [hunk 1]
```diff
@@ -467,6 +467,67 @@ def run_test (self):
         import_pub2 = self.nodes[0].getaddressinfo(addr2)['pubkey']
         assert_equal(pub2, import_pub2)
 
+        # Import some public keys to the keypool of a no privkey wallet
```
Comment (promag) [inline test/functional/wallet_importmulti.py]: Commit "Fetch keys from keypool when private keys are disabled"

Could move test to a different commit?

File: test/functional/wallet_importmulti.py [hunk 2]
```diff
@@ -467,6 +467,67 @@ def run_test (self):
...
+        self.log.info("Adding pubkey to keypool of disableprivkey wallet")
+        self.nodes[1].createwallet("noprivkeys", True)
```
Comment (promag) [inline test/functional/wallet_importmulti.py]: Commit "Fetch keys from keypool when private keys are disabled"

Use named argument `disable_private_keys=True`?

File: src/wallet/rpcdump.cpp [hunk 3]
```diff
@@ -1198,6 +1219,7 @@ UniValue importmulti(const JSONRPCRequest& mainRequest)
             "      \"internal\": <true>                                    , (boolean, optional, default: false) Stating whether matching outputs should be treated as not incoming payments\n"
             "      \"watchonly\": <true>                                   , (boolean, optional, default: false) Stating whether matching outputs should be considered watched even when they're not spendable, only allowed if keys are empty\n"
             "      \"label\": <label>                                      , (string, optional, default: '') Label to assign to the address (aka account name, for now), only allowed with internal=false\n"
+            "      \"keypool\": <true|false>                               , (boolean, optional, default: true) If true, adds the pubkeys to the keypool if private keys are disabled for the wallet.\n"
```
Comment (promag) [inline src/wallet/rpcdump.cpp]: Commit "Add option to importmulti add an imported pubkey to the keypool"

Default false.

File: src/wallet/rpcdump.cpp [hunk 4]
```diff
@@ -876,6 +877,16 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
...
         }
 
+        // Add to keypool only works with privkeys disabled
+        if (add_keypool && !pwallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Keys can only be imported to the keypool when private keys are disabled");
+        }
+
+        // Add to keypool only works with pubkeys
+        if (add_keypool && pubKeys.size() == 0) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Only pubkeys be imported to the keypool");
```
Comment (promag) [inline src/wallet/rpcdump.cpp]: Commit "Add option to importmulti add an imported pubkey to the keypool"

... can be ...

File: src/qt/walletmodel.cpp [hunk 5]
```diff
@@ -559,7 +559,7 @@ bool WalletModel::isWalletEnabled()
 
 bool WalletModel::privateKeysDisabled() const
 {
-    return m_wallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS);
+    return m_wallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) && m_wallet->KeypoolCountExternalKeys() == 0;
```
Comment (instagibbs) [inline src/qt/walletmodel.cpp]: This function should be renamed or called into from a new one.

`WalletModel::KeypoolEmptyAndPrivkeysDisabled` or something obnoxiously explicit.

File: src/wallet/rpcwallet.cpp [hunk 6]
```diff
@@ -161,11 +161,15 @@ static UniValue getnewaddress(const JSONRPCRequest& request)
...
-    if (pwallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)) {
-        throw JSONRPCError(RPC_WALLET_ERROR, "Error: Private keys are disabled for this wallet");
+    LOCK(pwallet->cs_wallet);
+
+    if (pwallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) && pwallet->KeypoolCountExternalKeys() == 0) {
+        throw JSONRPCError(RPC_WALLET_ERROR, "Error: Private keys are disabled for this wallet and no keys in the external keypool");
     }
 
-    LOCK(pwallet->cs_wallet);
+    if (!pwallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) && !pwallet->IsLocked()) {
```
Comment (instagibbs) [inline src/wallet/rpcwallet.cpp]: Might be time to make a `TryToTopUpKeyPool` function for when we will continue regardless. We do this in a number of places in the PR and probably elsewhere.

File: src/wallet/wallet.cpp [hunk 7]
```diff
@@ -3417,7 +3423,7 @@ bool CWallet::ReserveKeyFromKeyPool(int64_t& nIndex, CKeyPool& keypool, bool fRe
         if (!IsLocked())
             TopUpKeyPool();
 
-        bool fReturningInternal = IsHDEnabled() && CanSupportFeature(FEATURE_HD_SPLIT) && fRequestedInternal;
+        bool fReturningInternal = ((IsHDEnabled() && CanSupportFeature(FEATURE_HD_SPLIT)) || IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS))  && fRequestedInternal;
```
Comment (instagibbs) [inline src/wallet/wallet.cpp]: assumption here being using importmulti means you want to support hdsplit?

Perhaps break this into two lines:
```
bool fReturningInternal = fRequestedInternal;
fReturningInternal &= (IsHDEnabled() && CanSupportFeature(FEATURE_HD_SPLIT)) || IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS);
```

File: src/wallet/wallet.cpp [hunk 8]
```diff
@@ -3434,7 +3440,7 @@ bool CWallet::ReserveKeyFromKeyPool(int64_t& nIndex, CKeyPool& keypool, bool fRe
         if (!batch.ReadPool(nIndex, keypool)) {
             throw std::runtime_error(std::string(__func__) + ": read failed");
         }
-        if (!HaveKey(keypool.vchPubKey.GetID())) {
+        if (!IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) && !HaveKey(keypool.vchPubKey.GetID())) {
```
Comment (instagibbs) [inline src/wallet/wallet.cpp]: Belt and suspenders, we should be checking that this exists in our wallet regardless yes? Just not asking if we have the private key itself.

File: src/wallet/wallet.cpp [hunk 9]
```diff
@@ -3478,23 +3484,21 @@ void CWallet::ReturnKey(int64_t nIndex, bool fInternal, const CPubKey& pubkey)
 
 bool CWallet::GetKeyFromPool(CPubKey& result, bool internal)
 {
-    if (IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)) {
+    LOCK(cs_wallet);
+    if (IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) && KeypoolCountExternalKeys() == 0) {
```
Comment (instagibbs) [inline src/wallet/wallet.cpp]: This should be checking the keypool corresponding to `internal`, yes?

File: test/functional/wallet_importmulti.py (hunk 1 above)
Comment (achow101) [reply test/functional/wallet_importmulti.py]: Done

File: test/functional/wallet_importmulti.py (hunk 2 above)
Comment (achow101) [reply test/functional/wallet_importmulti.py]: Done

File: src/wallet/rpcdump.cpp (hunk 3 above)
Comment (achow101) [reply src/wallet/rpcdump.cpp]: Fixed

File: src/wallet/rpcdump.cpp (hunk 4 above)
Comment (achow101) [reply src/wallet/rpcdump.cpp]: Fixed

File: src/qt/walletmodel.cpp (hunk 5 above)
Comment (achow101) [reply src/qt/walletmodel.cpp]: Done

File: src/wallet/wallet.cpp (hunk 7 above)
Comment (achow101) [reply src/wallet/wallet.cpp]: Done

File: src/wallet/wallet.cpp (hunk 8 above)
Comment (achow101) [reply src/wallet/wallet.cpp]: I changed this line to check for the pubkey in the wallet instead of the privkey.

File: src/wallet/wallet.cpp (hunk 9 above)
Comment (achow101) [reply src/wallet/wallet.cpp]: Fixed

**Comment by achow101** [comment]
Rebased. The first commit is still from #14019

File: src/interfaces/wallet.h [hunk 10]
```diff
@@ -272,6 +272,9 @@ class Wallet
     //! Register handler for watchonly changed messages.
     using WatchOnlyChangedFn = std::function<void(bool have_watch_only)>;
     virtual std::unique_ptr<Handler> handleWatchOnlyChanged(WatchOnlyChangedFn fn) = 0;
+
+    // Get the number of external keypool keys
```
Comment (instagibbs) [inline src/interfaces/wallet.h]: generalize comment

File: src/wallet/wallet.cpp [hunk 11]
```diff
@@ -3311,6 +3311,22 @@ size_t CWallet::KeypoolCountExternalKeys()
...
+bool CWallet::NoPrivkeysAndKeypoolEmpty(bool internal)
+{
+    if (!IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)) {
+        return false;
+    }
+
+    {
+        LOCK(cs_wallet);
+        if (internal) {
+            return setInternalKeyPool.size() == 0;
```
Comment (instagibbs) [inline src/wallet/wallet.cpp]: mu-nit: add commit for KeypoolCountInternalKeys

**Comment by instagibbs** [comment]
`CWallet::CreateTransaction` needs to check for keys in change keypool even if `WALLET_FLAG_DISABLE_PRIVATE_KEYS` is set.

File: src/wallet/wallet.cpp (hunk 11 above)
Comment (achow101) [reply src/wallet/wallet.cpp]: I think that's unnecssary.

File: src/interfaces/wallet.h (hunk 10 above)
Comment (achow101) [reply src/interfaces/wallet.h]: Done

File: test/functional/wallet_importmulti.py [hunk 12]
```diff
@@ -446,6 +446,88 @@ def run_test (self):
...
+        # Importing multiple public keys with 2-of-2 p2sh multisig
+        self.log.info("Pubkeys should be available after importing 2-of-2 p2sh multisig with pubkeys")
+        addr1 = self.nodes[1].getnewaddress()
+        addr2 = self.nodes[1].getnewaddress()
+        pub1 = self.nodes[1].getaddressinfo(addr1)['pubkey']
+        pub2 = self.nodes[1].getaddressinfo(addr2)['pubkey']
+        ms = self.nodes[1].createmultisig(2, [pub1, pub2])
+        result = self.nodes[0].importmulti(
+            [{
+                'scriptPubKey' : { 'address' : ms['address']},
```
Comment (practicalswift) [inline test/functional/wallet_importmulti.py]: FWIW, PEP-8 whitespace nits:

```
./test/functional/wallet_importmulti.py:458:31: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:458:35: E201 whitespace after '{'
./test/functional/wallet_importmulti.py:458:45: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:459:31: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:460:26: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:482:31: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:482:35: E201 whitespace after '{'
./test/functional/wallet_importmulti.py:482:45: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:483:31: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:484:26: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:485:26: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:504:31: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:504:35: E201 whitespace after '{'
./test/functional/wallet_importmulti.py:504:45: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:505:31: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:506:26: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:507:26: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:508:27: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:523:27: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:523:31: E201 whitespace after '{'
./test/functional/wallet_importmulti.py:523:41: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:524:27: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:525:22: E203 whitespace before ':'
./test/functional/wallet_importmulti.py:526:22: E203 whitespace before ':'
```

**Comment by Sjors** [comment]
Concept ACK. Lightly tested via your combined branch 3fa968ea343: I was able to import a bunch of receive and change keys. I was also able to receive coins on a `bech32` address and send from it.

For some reason when I composed the transaction it picked change address with index 1 instead of with index 0. Is that an existing rule?

What I also found confusing, though I don't know if that's this PR, or the other two or just existing weirdness, is how `dumpwallet` shows `p2sh-p2wpkh` addresses under `# addr`, even though I launched bitcoind with `-addresstype=bech32`.

**Comment by achow101** [comment]
> For some reason when I composed the transaction it picked change address with index 1 instead of with index 0. Is that an existing rule?

Maybe your import command was weird? It should be adding them in the order of the import command so you should be getting them in that order too.

File: test/functional/wallet_importmulti.py (hunk 12 above)
Comment (achow101) [reply test/functional/wallet_importmulti.py]: Fixed

File: src/wallet/rpcdump.cpp [hunk 13]
```diff
@@ -917,6 +928,58 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
...
 
+            // Import public keys
+            for (size_t i = 0; i < pubKeys.size(); ++i) {
+                const std::string& pubkey = pubKeys[i].get_str();
+
+                if (!IsHex(pubkey)) {
+                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Pubkey must be a hex string");
+                }
+
+                std::vector<unsigned char> vData(ParseHex(pubkey));
```
Comment (practicalswift) [inline src/wallet/rpcdump.cpp]: This `vData` shadows another local variable `vData` :-)

**Comment by Sjors** [comment]
What's a good way to inspect the wallet / keypool?

**Comment by achow101** [comment]
@Sjors You can use dumpwallet

**Comment by Sjors** [comment]
@achow101 `dumpwallet` doesn't contain the bip32 derivation paths though.

**Comment by achow101** [comment]
I will rework this to not require #14019 as soon as I have time (in a few days)

**Comment by Sjors** [comment]
Probably best rebased on #14491 or its predecessor.

**Comment by achow101** [comment]
Rebased onto #14491

**Comment by achow101** [comment]
I've added two commits contributed by @instagibbs. The first one makes it so that things are imported in the order they are listed in the import command or in the order expanded for ranged descriptors. The second commit has watch keys imported to the keypool be marked as used when transactions are seen using them.

File: src/wallet/rpcdump.cpp [hunk 14]
```diff
@@ -804,233 +805,368 @@ UniValue dumpwallet(const JSONRPCRequest& request)
...
-                if (pubkey.size() && pubkey_temp != pubkey) {
-                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Private key does not match public key for address");
+        if (!error.empty()) {
+            warnings.push_back("Ignoring information in keys, pubkeys, redeemscript, and witnessscript:: " + error);
+            import_data = ImportData();
+        } else {
+            // RecurseImportData() removes any relevant redeemscript/witnessscript from import_data, so we can use that to discover if a superfluous one was provided.
+            if (import_data.redeemscript) warnings.push_back("Ignoring redeemscript " + strRedeemScript + " as this is not a P2SH script.");
+            if (import_data.witnessscript) warnings.push_back("Ignoring witnessscript " + witness_script_hex + " as this is not a P2WSH script.");
+            for (const auto& privkey : import_data.privkeys) {
```
Comment (practicalswift) [inline src/wallet/rpcdump.cpp]: Could use `std::any_of`? :-)

**Comment by MeshCollider** [comment]
Light utACK https://github.com/bitcoin/bitcoin/pull/14075/commits/8e6b7bd5e9aeb8efe85068de8fc38cf0b22f1185

Will review more in-depth after the dependent PRs are merged and this is rebased

**Comment by Sjors** [comment]
I think we should make `keypoolrefill` and `TopUpKeyPool` work with just public keys.

When `WALLET_FLAG_DISABLE_PRIVATE_KEYS` is set we can skip the `EnsureWalletIsUnlocked(pwallet)` check.

`TopUpKeyPool` currently only creates random fresh keys (for legacy wallets) or derives new keys from the wallet HD seed. We could change it to first walk through `solvable` keys in the wallet.

For safety we could prevent it from putting keys without private keys into the keypool for wallets without `WALLET_FLAG_DISABLE_PRIVATE_KEYS`.

In that case we don't need the `keypool` option in `importmulti`.

This might make it easier to refactor towards a descriptor based wallet without a keypool, because we avoid adding another code path where stuff can be added to the keypool. In such a refactor a RPC method like `getnewaddress` will have to reference a descriptor. That would be the wallet receive descriptor by default. `importmulti` (or something like it) would be then be able to set the wallet receive descriptor (in case the wallet is empty).

**Comment by achow101** [comment]
Rebased.

I dropped the last commit as #14821 should handle that.

**Comment by achow101** [comment]
I've added a commit which fixed an issue where the `Request payment` button was not being enabled once keys were imported to the keypool. That commit sends a signal to the GUI every time the keypool changes so that the button will be enabled and disabled as there keys are added and removed from the keypool.

**Comment by achow101** [comment]
I split out 28c89d1 into it's own PR: #15225

**Comment by Sjors** [comment]
In light of #15226 I'd like to revisit my [earlier remark](
https://github.com/bitcoin/bitcoin/pull/14075#issuecomment-446547958):

I'd like to avoid messing with the keypool inside of `importmulti`. Instead I would prefer if `TopUpKeyPool` can deal with imported keys.

> `TopUpKeyPool` currently only creates random fresh keys (for legacy wallets) or derives new keys from the wallet HD seed. We could change it to first walk through solvable keys in the wallet.

> In that case we don't need the `keypool` option in `importmulti`.

**Comment by jonasschnelli** [comment]
Concept ACK

Ideally – but orthogonal to this PR – would be if someone could provide a xpub or range based descriptor during sethdseed (or new command) that would make the keypool act similar to the default HD keypool but derive watch only pub keys.

**Comment by Sjors** [comment]
Tested a4bd25d, import works and GUI new address button updates as expected.

**Comment by Sjors** [comment]
I had some difficulty combining this with #15226 (blank wallet) and made a commit that does that, poorly: https://github.com/Sjors/bitcoin/commit/8988a96e2eff8bd6344f9e93032e86742e7dc3f8
However, the blank wallet PR has since changed again, so my commit is merely useful for inspiration and some extra tests. Don't let it hold back merging here.

**Comment by MeshCollider** [comment]
I don't think we can get both this and #14021 in to 0.18 so I'll remove this from the milestone

**Comment by achow101** [comment]
I've actually decided to base this on #15226. In the version of #15226, I introduced `CanGetAddresses` and `CanGenerateKeys` which I am using here for determining whether there are things in the keypool to fetch keys instead of `NoPrivkeysAndKeypoolEmpty` (which did a similar thing but was less robust). Because of this, I have also replaced `WalletModel::privateKeysDisabled` with `canGetaddresses`.

File: src/qt/overviewpage.cpp [hunk 15]
```diff
@@ -184,7 +184,7 @@ void OverviewPage::setBalance(const interfaces::WalletBalances& balances)
     // for symmetry reasons also show immature label when the watch-only one is shown
     ui->labelImmature->setVisible(showImmature || showWatchOnlyImmature);
     ui->labelImmatureText->setVisible(showImmature || showWatchOnlyImmature);
-    ui->labelWatchImmature->setVisible(!walletModel->privateKeysDisabled() && showWatchOnlyImmature); // show watch-only immature balance
+    ui->labelWatchImmature->setVisible(walletModel->canGetAddresses() && showWatchOnlyImmature); // show watch-only immature balance
```
Comment (Sjors) [inline src/qt/overviewpage.cpp]: Same as above, but even more rare.

File: src/wallet/wallet.cpp [hunk 16]
```diff
@@ -2756,7 +2756,7 @@ bool CWallet::CreateTransaction(interfaces::Chain::Lock& locked_chain, const std
                 //  post-backup change.
 
                 // Reserve a new key pair from key pool
-                if (IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)) {
+                if (!CanGetAddresses(true)) {
                     strFailReason = _("Can't generate a change-address key. Private keys are disabled for this wallet.");
```
Comment (Sjors) [inline src/wallet/wallet.cpp]: Nit: drop / tweak "Private keys are disabled for this wallet."

File: src/wallet/wallet.cpp [hunk 17]
```diff
@@ -3417,7 +3417,8 @@ bool CWallet::ReserveKeyFromKeyPool(int64_t& nIndex, CKeyPool& keypool, bool fRe
         if (!IsLocked())
             TopUpKeyPool();
 
-        bool fReturningInternal = IsHDEnabled() && CanSupportFeature(FEATURE_HD_SPLIT) && fRequestedInternal;
+        bool fReturningInternal = fRequestedInternal;
+        fReturningInternal &= (IsHDEnabled() && CanSupportFeature(FEATURE_HD_SPLIT)) || IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS);
```
Comment (Sjors) [inline src/wallet/wallet.cpp]: nit, add `HasHDSeed()` to complement `IsHDEnabled()`, and make it more clear what you're checking:
```cpp
// All WALLET_FLAG_DISABLE_PRIVATE_KEYS wallets were created after HD_SPLIT was added.
const bool fSupportsInternal = (HasHDSeed() && IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)) || CanSupportFeature(FEATURE_HD_SPLIT)
bool fReturningInternal = fRequestedInternal & fSupportsInternal;
```

File: src/wallet/wallet.h [hunk 18]
```diff
@@ -993,6 +993,7 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
     bool NewKeyPool();
     size_t KeypoolCountExternalKeys() EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
     bool TopUpKeyPool(unsigned int kpSize = 0);
+    void AddKeypoolPubkey(const CPubKey& pubkey, const bool internal);
```
Comment (Sjors) [inline src/wallet/wallet.h]: Nit: call this new function from `TopUpKeyPool`, so it's more clear this function reuses existing code. It could take a `error_message` and an optional `WalletBatch` argument. If `WalletBatch` is set it should skip `NotifyCanGetAddressesChanged()`, though that gets a bit ugly.

**Review by Sjors — APPROVED** [review]
I'd still like to get this into 0.18 even if it's not tagged as such.

~tACK 8484a10~ (_see newer comment_) modulo `privateKeysDisabled` needs to be restored on the QT overview page (I can live with punting that).

For a followup I'd like to loosen the restriction of not adding imported private keys to the keypool. That's not necessary for [HWI](https://github.com/bitcoin-core/HWI), so can wait until after the branching. Implemented here: https://github.com/Sjors/bitcoin/commits/2019/02/add_privkey_to_keypool

Consider squashing dffff150e98fc96c98e9e8caf40cec7391bbebfb into 39c95977b2435110114cffd620a2efdb02357360

File: test/functional/wallet_importmulti.py [hunk 19]
```diff
@@ -625,6 +625,74 @@ def run_test(self):
...
+            },
+            {
+                'desc': 'wpkh(' + pub2 + ')',
+                'keypool': True,
+                'internal': True,
+                "timestamp": "now",
+            }]
+        )
+        assert result[0]['success']
+        assert result[1]['success']
```
[the remark this answers is not in the document]
Comment (achow101) [reply test/functional/wallet_importmulti.py]: Done

File: src/qt/bitcoingui.cpp [hunk 20]
```diff
@@ -1202,7 +1202,7 @@ void BitcoinGUI::updateWalletStatus()
     }
     WalletModel * const walletModel = walletView->getWalletModel();
     setEncryptionStatus(walletModel->getEncryptionStatus());
-    setHDStatus(walletModel->privateKeysDisabled(), walletModel->wallet().hdEnabled());
+    setHDStatus(!walletModel->canGetAddresses(), walletModel->wallet().hdEnabled());
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/qt/bitcoingui.cpp]: I've reverted this change.

File: src/qt/overviewpage.cpp [hunk 21]
```diff
@@ -161,7 +161,7 @@ void OverviewPage::setBalance(const interfaces::WalletBalances& balances)
 {
     int unit = walletModel->getOptionsModel()->getDisplayUnit();
     m_balances = balances;
-    if (walletModel->privateKeysDisabled()) {
+    if (!walletModel->canGetAddresses()) {
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/qt/overviewpage.cpp]: I reverted that change

File: src/qt/overviewpage.cpp (hunk 15 above)
Comment (achow101) [reply src/qt/overviewpage.cpp]: Reverted

File: src/wallet/wallet.h (hunk 18 above)
Comment (achow101) [reply src/wallet/wallet.h]: I added a function `AddKeypoolPubkeyWithDB` which `TopUpKeypool` calls. `AddKeypoolPubkey` will make the `WalletBatch`, call `AddKeypoolPubkeyWithDB` and then ``NotifyCanGetAddressesChanged``.

File: src/wallet/wallet.cpp (hunk 17 above)
Comment (achow101) [reply src/wallet/wallet.cpp]: We can do that later.

File: src/wallet/wallet.cpp (hunk 16 above)
Comment (achow101) [reply src/wallet/wallet.cpp]: Changed.

File: test/functional/wallet_importmulti.py [hunk 22]
```diff
@@ -625,6 +625,76 @@ def run_test(self):
                      ismine=False,
                      iswatchonly=False)
 
+        # Import some public keys to the keypool of a no privkey wallet
```
Comment (instagibbs) [inline test/functional/wallet_importmulti.py]: A test showing that importing some non-single-pubkey addresses using descriptors doesn't cause the keypool to grow seems apt.

Comment (achow101) [reply test/functional/wallet_importmulti.py]: Done

**Comment by Sjors** [comment]
Yikes, I missed something in my earlier tACK, because I was too focussed on testing with private key support.

When I create a watch-only wallet, import keys and rescan, it produces the right addresses, but it's not picking up existing transactions, not detecting new incoming transactions.

```
importmulti '[{"desc": "sh(wpkh(tpubD6NzVbkrYhZ4YNXVQbNhMK1WqguFsUXceaVJKbmno2aZ3B6QfbMeraaYvnBSGpV3vxLyTTK9DYT1yoEck4XUScMzXoQ2U2oSmE2JyMedq3H/0h/0h/*))", "range": {"end": 2}, "timestamp": 1548979200, "keypool": true, "watchonly": true}]
```

```
getaddressinfo tb1q8vprchan07gzagd5e6v9wd7azyucksq2y38hpw
{
  "address": "tb1q8vprchan07gzagd5e6v9wd7azyucksq2y38hpw",
  "scriptPubKey": "00143b023c5fb37f902ea1b4ce985737dd11398b400a",
  "ismine": false,
  "solvable": true,
  "desc": "wpkh([3b023c5f]021ec1fdc1d288da27c93a79b0fb7f669314a6f477400d2a1019529f7ca2ee27b4)",
  "iswatchonly": false,
  "isscript": false,
  "iswitness": true,
  "witness_version": 0,
  "witness_program": "3b023c5fb37f902ea1b4ce985737dd11398b400a",
  "pubkey": "021ec1fdc1d288da27c93a79b0fb7f669314a6f477400d2a1019529f7ca2ee27b4",
  "label": "",
  "ischange": false,
  "labels": [
    {
      "name": "",
      "purpose": "receive"
    }
  ]
}
```

This looks correct to me (calling `deriveaddresses` on that inferred descriptor returns the address).

Remember when @instagibbs added:

> The second commit has watch keys imported to the keypool be marked as used when transactions are seen using them.

You then dropped that commit, because it was superseded by #14821.

The problem is that `AddToWalletIfInvolvingMe` only considers `IsMine` https://github.com/bitcoin/bitcoin/blame/master/src/wallet/wallet.cpp#L972. It's possible that some earlier version of #14021 incorrectly set IsMine so we didn't notice this while testing `hww`.

**Comment by achow101** [comment]
> But these imported public keys are merely `Solvable` so why aren't they skipped?

`IsMine()` returns true for things that are watch only, solvable, or spendable.

**Comment by instagibbs** [comment]
More concretely, `IsMine` does a last-minute check if the exact scriptPubKey exists in the watchonly store if it fails at everything else.

**Comment by instagibbs** [comment]
Feel free to remove "built on" comment in OP so I don't keep checking if it's time to review in earnest :)

**Comment by instagibbs** [comment]
There is no test for keys being imported in order, and it appears to not actually be implemented for descriptors with ranges? You're iterating over a map's keys, which will be sorted in `CKeyID` order.

**Comment by instagibbs** [comment]
Good news: I'm wrong, and I'm the one who wrote the logic originally :( 

Test and I'm happy.

File: test/functional/wallet_importmulti.py [hunk 23]
```diff
@@ -625,6 +625,119 @@ def run_test(self):
...
+        addresses = [
+            'bcrt1qtmp74ayg7p24uslctssvjm06q5phz4yrxucgnv', # m/0'/0'/0
+            'bcrt1q8vprchan07gzagd5e6v9wd7azyucksq2xc76k8', # m/0'/0'/1
+            'bcrt1qtuqdtha7zmqgcrr26n2rqxztv5y8rafjp9lulu', # m/0'/0'/2
+            'bcrt1qau64272ymawq26t90md6an0ps99qkrse58m640', # m/0'/0'/3
+            'bcrt1qsg97266hrh6cpmutqen8s4s962aryy77jp0fg0', # m/0'/0'/4
+        ]
+        result = wrpc.importmulti(
+            [{
+                'desc': 'wpkh(' + xpub + '/0h/0h/*' + ')',
```
Comment (instagibbs) [inline test/functional/wallet_importmulti.py]:
```
./test/functional/wallet_importmulti.py:730:35: F821 undefined name 'xpub'
```

**Comment by Sjors** [comment]
@instagibbs keys from a ranged descriptor are imported in order. Feel free to replace this test:

```py
        # Test importing of a ranged descriptor without keys, check order
        self.log.info("Should import the ranged descriptor with specified range as solvable")
        self.test_importmulti({"desc": desc,
                               "timestamp": "now",
                               "range": {"end": 9}},
                              success=True,
                              warnings=["Some private keys are missing, outputs will be considered watchonly. If this is intentional, specify the watchonly flag."])
        # Check the first two addresses:
        for address in addresses:
            test_address(self.nodes[1],
                         key.p2sh_p2wpkh_addr,
                         solvable=True)

        # Check keys were imported in the correct order:
        for i in range(10):
            address = self.nodes[1].getnewaddress()
            assert_equal(self.nodes[1].getaddressinfo(address)["hdkeypath"], "m/0'/0'/%s'" % i)
```

File: test/functional/wallet_importmulti.py (hunk 23 above)
Comment (achow101) [reply test/functional/wallet_importmulti.py]: Whoops

**Comment by instagibbs** [comment]
Test fix only change.

reACK https://github.com/bitcoin/bitcoin/pull/14075/commits/32fd3bb627fbd3c44b4f45434ead56f207dffa06

**Comment by instagibbs** [comment]
re-ACK https://github.com/bitcoin/bitcoin/pull/14075/commits/38fbc101d1342222565b8670b7cd9c2485a57c74

Ordering test fixed, at least locally.

**Comment by Sjors** [comment]
utACK a6e2c3331023d11112f8e9d176aa823d84b3f5ce (only adds a test since my last ACK)

**Comment by MeshCollider** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/14075/commits/38fbc101d1342222565b8670b7cd9c2485a57c74

**Comment by MeshCollider** [comment]
LGTM https://github.com/bitcoin/bitcoin/pull/14075/commits/f4b00b70e811d3772589ab3c64d7658f4dbdab69

**Comment by Sjors** [comment]
Post merge utACK f4b00b7
